### DISCUSSION TITLE
Aws restJson1 Protocol and Compliance Tests - Lets Discuss

### DISCUSSION THREAD

**Discussion by yisraelU** [disneystreaming/smithy4s#702] [General]
A couple of features that we don't support , that are supported by AWS and are tested in the AWS Compliance tests
1. Infinity,Negative Infinity, NaN as values for Float/Double
2. mediaType trait

**Comment by Baccata** [comment]
We do support the `jsonName` trait, can you be more specific as to the context where you've noticed it's not supported ?

**Reply by kubukoz** [reply]
infinity/negative infinity are supported in scala's double too, and `@mediaType` is supported to some extent AFAIK. I'd suggest reproducing these in some tests.

**Reply by yisraelU** [reply]
double and float , however the fact that Scala supports them is irrelevant (I mean it is a necessary part , but its not the cause of the problem)
1. Smithy defines them as Strings which results in a Document.DString and will fail when looking for a float
```params: {
            queryFloat: "Infinity",
            queryDouble: "Infinity",
            queryParamsMapOfStringList: {
                "Float": ["Infinity"],
                "Double": ["Infinity"],
            }
        }
```


2.  we wrap everything in BigDecimal which does not support those and will throw an exception with the following code
```
val nan = java.lang.Double.NaN
val bigD = BigDecimal(nan)
```

**Comment by yisraelU** [comment]
Ok for starters I assumed jsonName is not supported as it doesnt appear in this list https://disneystreaming.github.io/smithy4s/docs/protocols/simple-rest-json/overview
I also did a search for it on the website and could not find a mention.
that said , every single compliance test based on jsonName failed

**Reply by kubukoz** [reply]
Could it have been an issue of a misconfigured hint mask?

**Reply by kubukoz** [reply]
> Ok for starters I assumed jsonName is not supported as it doesnt appear in this list [disneystreaming.github.io/smithy4s/docs/protocols/simple-rest-json/overview](https://disneystreaming.github.io/smithy4s/docs/protocols/simple-rest-json/overview)

it does though:

<img width="603" alt="image" src="https://user-images.githubusercontent.com/894884/210421504-b003f775-8af9-4aa6-88a0-646e4b653a39.png">

what's wrong with the docs is that there are two "Supported traits" headings, and the first one (below) lists **shapes** _and_ traits.

<img width="988" alt="image" src="https://user-images.githubusercontent.com/894884/210421595-bde6282d-64bb-44ac-8f23-ca9e10ea5701.png">

**Reply by yisraelU** [reply]
i will definitely check , I just assumed we didn't support it . Either way documentation should be updated or point to Alloy

**Reply by yisraelU** [reply]
Ahh... ok I yield

**Reply by yisraelU** [reply]
now I need to look into failures

**Reply by yisraelU** [reply]
actually.. @kubukoz  jsonName is not supported on Structures which explains all the failures ,its only supported on Unions at least from my read of the SchemaVisitorJCodec

**Reply by Baccata** [reply]
@yisraelU `jsonName` is supported on structures fields : https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/json/src/smithy4s/http/json/SchemaVisitorJCodec.scala#L908-L912

For the sake of the discussion, I'll ask you (and everyone getting involved) to apply the following approach : 

1. If you think something is not supported when it should, please share a repro (by means of a scala-cli script or a PR against this repro adding a unit-test showcasing the problem) 
2. If you don't manage to produce a unit-test repro but the compliance-tests are failing, it may be that there is a gap in the compliance-test implementation (as opposed to the protocol implementation), please add a link to the test in question (and if you can, a repro invoking the compliance test by means of a scala-cli script)
3. If you see a gap in the documentation where we have forgotten to indicate that something is supported, please link to the paragraph (or share a screenshot similar to what @kubukoz did). 

It's important to abide by this approach for various reasons : 

* first of all, it forces people through a rubber-duck process : they might come to some realisations on their own, or learn some things about the implementation during the reproduction of the problem 
* it really smoothens communication : as a maintainer, having your hands on a reproduction (rather than a subjective description of the issue) is a huge time saver and reduces risks of misunderstanding and mismatches of semantics between the interlocutors 
* it helps people gain experience and teaches them what to expect from contributors if they ever end up becoming main maintainer of an open-source project

**Reply by yisraelU** [reply]
@Baccata Affirmative to everything you wrote.
And I think you meant this one https://github.com/disneystreaming/smithy4s/blob/4c9a81bfa609ef4a8adadccac7694a496882ec60/modules/json/src/smithy4s/http/json/SchemaVisitorJCodec.scala#L1060

**Reply by Baccata** [reply]
Indeed I did

**Reply by yisraelU** [reply]
I believe its more about how we test for compliance . Ill open a repr to showcase

**Reply by yisraelU** [reply]
https://github.com/disneystreaming/smithy4s/pull/704

**Comment by yisraelU** [comment]
My understanding is the default serialization for timestamps in a json under the aws restJson1 protocol is [epoch seconds](https://smithy.io/2.0/aws/protocols/aws-restjson1-protocol.html?highlight=restjson#json-shape-serialization) as opposed to simpleRestJson is date time
